### DISCUSSION TITLE
How to achieve real-time ray tracing?

### DISCUSSION THREAD

**Discussion by qian-o** [dotnet/Silk.NET#1870] [Q&A] [answered]
I have been using OpenGL because it is simple. Recently, I wanted to implement ray tracing in my program, but GL only has a compute pipeline and not a ray tracing pipeline.
If I switch to Vulkan, it will not only require a significant amount of work but also have a steep learning curve. So, are there any ready-made SDKs or better ways in .NET to achieve this?

**Comment by Perksey** [comment, marked as answer]
No, you will need to switch to Vulkan to use VK_KHR_ray_tracing or switch to Direct3D 12 to use DXR. You may not need to switch your entire program over, you could interoperate with those APIs, but it’s far easier to just use Vulkan through and through. There is no way outside of interoperation with another graphics API to use ray tracing with OpenGL.

**Reply by qian-o** [reply]
Thank you, that's actually a great suggestion. Do have any recommendations for Vulkan tutorials or books? I've tried following the 'vulkan-tutorial' for some basic content, but I feel like it's not what I'm looking for.

**Reply by Perksey** [reply]
vkguide.dev is good
